url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/core/NetBase.java

```diff
@@ -0,0 +1,796 @@
+ * <br></br>
+ * @see Net
+ * @author Sebastian Otte
+ */
+public abstract class NetBase implements Net, Serializable, Cloneable {
+    private static final long serialVersionUID = 4485482287977886674L;
+    
+    /**
+     * Gives the standard lower bound for random weight initialization.
+     */
+    public static final double RANDOM_WEIGHT_LBD = -0.1;
+    /**
+     * Gives the standard upper bound for random weight initialization.
+     */
+    public static final double RANDOM_WEIGHT_UBD = 0.1;
+    /**
+     * Provides the network structure of this network instance.
+     * @see NetStructure
+     */
+    protected NetStructure structure = null;
+    /**
+     * Provides the network data of this network instance.
+     */
+    protected NetData data = null;
+    /**
+     * Gives the current frameidx of the network. This can be
+     * considered as an inner state during the computation process.
+     */
+    protected int frameidx = 0;
+    /**
+     * Provides a read port to the output data of this network. This port 
+     * targets to the range of output cells in the output buffer 
+     * using the current frameidx.
+     */
+    private ReadPort outputport = null;
+    /**
+     * Provides a write port to the input data of this network. This port
+     * targets to the range of input cells in the output buffer
+     * using the current frameidx. 
+     */
+    private WritePort inputport = null;
+    /**
+     * Provides a write port to the target data of this network (necessary
+     * to compute the error). This port targets to the range of output cells
+     * the gradinput buffer.
+     */
+    private WritePort targetport = null;
+    /**
+     * This methods computes the activation (also integration) of a single 
+     * layer given by idx, based on the current frameidx. The method 
+     * ensures that the activations of all arrays of the layer are computed 
+     * in correct time order, which has been defined during generating the network.
+     * <br></br>
+     * @param idx Gives a layer index.
+     */
+    protected void computeLayerActivations(final int idx) {
+        final Layer layer = this.structure.layers[idx];
+        //
+        // respect shifted computation indices.
+        //
+        for (int c = 0; c < layer.compwidth; c++) {
+            //
+            final int lbd = layer.complbds[c];
+            final int ubd = layer.compubds[c];
+            //
+            // integration.
+            //
+            for (int a = lbd; a <= ubd; a++) {
+                final CellArray array = this.structure.arrays[a];
+                //
+                CellIntegration.perform(
+                    this.data.output[this.frameidx], this.data.input[this.frameidx], 
+                    array.cellslbd, array.cellsnum,
+                    this.data.weights, this.structure.links,
+                    array.predslbd, array.predsnum,
+                    array.celltype.integration
+                );
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 105:115, 152:170.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.